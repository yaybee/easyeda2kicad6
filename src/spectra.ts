export interface ISpectraList extends Array<ISpectraList | string | number | null> {}

const WHITESPACE = [' ', '\t', '\r', '\n'];

function notNull<TValue>(value: TValue | null): value is TValue {
  return value !== null;
}

export function encodeString(str: string) {
  // output formatting
  if (str === '_LF_') {
    return '\n';
  }
  if (str === '_LF1_') {
    return '\n  ';
  }
  if (str === '_LF2_') {
    return '\n    ';
  }
  if (str === '_LF3_') {
    return '\n      ';
  }
  if (str === '_LF4_') {
    return '\n        ';
  }
  // make sting without quotes for eg keywords
  if (/^[a-z][a-z0-9_]+$/.test(str)) {
    return str;
    // make special quoted string possible by starting char &
  } else if (/^&/.test(str)) {
    return `"${str.replace('&', '')}"`;
  }
  // make quoted string eg for values
  // "" in text field throws Kicad error: changed in ' (BUG FIX)
  return `"${str.replace(/"/g, `'`)}"`;
}

function encodeNumber(value: number) {
  return (Math.round(value * 1000 + Number.EPSILON) / 1000).toString();
}

export function encodeValue(value: ISpectraList | string | number) {
  if (typeof value === 'string') {
    return encodeString(value);
  }
  if (typeof value === 'number') {
    return encodeNumber(value);
  }
  return encodeObject(value);
}

export function encodeObject(object: ISpectraList): string {
  return '(' + object.filter(notNull).map(encodeValue).join(' ') + ')';
}

function parseElement(input: string): [ISpectraList | string | number, number] {
  let idx = 0;
  while (WHITESPACE.includes(input[idx])) {
    idx++;
  }
  if (idx >= input.length) {
    throw new Error('Unexpected end of string');
  }
  if (input[idx] === '(') {
    idx++;
    const result = [];
    while (input[idx] !== ')') {
      if (idx >= input.length) {
        throw new Error('Unexpected end of string');
      }
      const [element, len] = parseElement(input.substr(idx));
      result.push(element);
      idx += len;
    }
    return [result, idx + 1];
  } else if (input[idx] === '"') {
    idx++;
    let result = '';
    while (input[idx] !== '"') {
      result += input[idx];
      idx++;
      if (input.substr(idx, 2) === '""') {
        result += '"';
        idx += 2;
      }
    }
    return [result, idx + 1];
  } else {
    let result = '';
    while (![...WHITESPACE, '(', ')'].includes(input[idx])) {
      if (idx >= input.length) {
        throw new Error('Unexpected end of string');
      }
      result += input[idx];
      idx++;
    }
    const numVal = parseFloat(result);
    if (typeof numVal === 'number' && !isNaN(numVal)) {
      return [numVal, idx];
    } else {
      return [result, idx];
    }
  }
}

export function parseObject(spectra: string) {
  spectra = spectra.trim();
  const [parsed] = parseElement(spectra);
  return parsed;
}
