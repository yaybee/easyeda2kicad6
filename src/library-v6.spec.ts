import * as fixtureJSON from './fixtures/testingLibrary-v6.json';
import { convertLibrary, convertPin, convertText, IProperties } from './library-v6';
import { IConversionState } from './schematic-v6';

import { ISpectraList } from './spectra';

function removeNullsAndFormating(a: ISpectraList): ISpectraList {
  return a
    .map((item) => (item instanceof Array ? removeNullsAndFormating(item) : item))
    .filter((e) => e !== null)
    .filter((e) => e !== '_LF_')
    .filter((e) => e !== '_LF1_')
    .filter((e) => e !== '_LF2_')
    .filter((e) => e !== '_LF3_')
    .filter((e) => e !== '_LF4_');
}

function round(obj: ISpectraList | string | number, precision = 3): ISpectraList | string | number {
  if (obj instanceof Array) {
    return obj.map((item) => round(item, precision));
  }
  if (typeof obj === 'number') {
    const result = parseFloat(obj.toFixed(precision));
    if (result > -Number.EPSILON && result < Number.EPSILON) {
      return 0;
    }
    return result;
  }
  return obj;
}

function normalize(obj: ISpectraList) {
  return round(removeNullsAndFormating(obj));
}

const compProp: IProperties = {
  ref: '',
  value: '',
  pre: '',
  lib: '',
  rotation: 0,
  package: '',
  pinNameShowCount: 0,
  pinNameHideCount: 0,
  pinNumberShowCount: 0,
  pinNumberHideCount: 0,
  component: [],
};
const conversionState: IConversionState = {
  schRepCnt: 0,
  schReports: [],
  schReportsPosition: 0,
  libTypes: {},
  savedLibs: [],
  savedLibMsgs: [],
  convertingSymFile: true,
};
const transform = { x: 400, y: 300 };
const CRCTable: number[] = [];

function convertShape(shape: string) {
  const [type, ...args] = shape.split('~');
  switch (type) {
    case 'P':
      return convertPin(args, transform, compProp, conversionState);
    case 'T':
      return convertText(args, transform);
    default:
      console.warn(`Warning: unsupported shape ${type}`);
      return null;
  }
}
describe('convertShape as standin for conversions from function convertLibrary', () => {
  it('should convert P (pin) into Kicad pin', () => {
    const input =
      'P~show~3~3~450~310~0~gge25~0^^450~310^^M 450 310 h -14~#880000^^0~428~313~0~VCC~end~~5.5pt~#0000FF^^1~435~309~0~3~start~~5.5pt~#0000FF^^1~433~310^^0~M 430 307 L 427 310 L 430 313';
    expect(normalize(convertShape(input))).toEqual([
      [
        'pin',
        'bidirectional',
        'inverted',
        ['at', 12.7, -2.54, 180],
        ['length', 5.08],
        ['name', 'VCC', ['effects', ['font', ['size', 1.27, 1.27]]]],
        ['number', '3', ['effects', ['font', ['size', 1.27, 1.27]]]],
      ],
    ]);
  });
});

describe('convertLibrary full test', () => {
  it('should convert an EasyEDA library file to a Kicad symbol file', () => {
    const schRes = convertLibrary(null, fixtureJSON, conversionState, CRCTable, 'test');
    expect(schRes).toMatchSnapshot();
  });
});
