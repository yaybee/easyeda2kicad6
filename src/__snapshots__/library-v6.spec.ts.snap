// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertLibrary full test should convert an EasyEDA library file to a Kicad symbol file 1`] = `
Array [
  "_LF1_",
  Array [
    "symbol",
    "test:PC357NJ0000F",
    null,
    Array [
      "pin_names",
      Array [
        "offset",
        2,
      ],
      "hide",
    ],
    Array [
      "in_bom",
      "yes",
    ],
    Array [
      "on_board",
      "yes",
    ],
    "_LF2_",
    Array [
      "property",
      "Reference",
      "U",
      Array [
        "id",
        0,
      ],
      Array [
        "at",
        0,
        0,
        0,
      ],
      Array [
        "effects",
        Array [
          "font",
          Array [
            "size",
            1.27,
            1.27,
          ],
        ],
        null,
      ],
    ],
    "_LF2_",
    Array [
      "property",
      "Value",
      "PC357NJ0000F",
      Array [
        "id",
        1,
      ],
      Array [
        "at",
        0,
        0,
        0,
      ],
      Array [
        "effects",
        Array [
          "font",
          Array [
            "size",
            1.27,
            1.27,
          ],
        ],
        null,
      ],
    ],
    "_LF2_",
    Array [
      "property",
      "Footprint",
      "OPTO-SMD-4_L4.4-W4.1-P2.54-LS7.0-BR",
      Array [
        "id",
        2,
      ],
      Array [
        "at",
        0,
        0,
        0,
      ],
      Array [
        "effects",
        Array [
          "font",
          Array [
            "size",
            1.27,
            1.27,
          ],
        ],
        "hide",
      ],
    ],
    "_LF2_",
    Array [
      "property",
      "ki_keywords",
      "PC357NJ0000F",
      Array [
        "id",
        4,
      ],
      Array [
        "at",
        0,
        0,
        0,
      ],
      Array [
        "effects",
        Array [
          "font",
          Array [
            "size",
            1.27,
            1.27,
          ],
        ],
        "hide",
      ],
    ],
    "_LF2_",
    Array [
      "property",
      "ki_description",
      "part manufactured by: Sharp Microelectronics",
      Array [
        "id",
        5,
      ],
      Array [
        "at",
        0,
        0,
        0,
      ],
      Array [
        "effects",
        Array [
          "font",
          Array [
            "size",
            1.27,
            1.27,
          ],
        ],
        "hide",
      ],
    ],
    "_LF2_",
    Array [
      "symbol",
      "PC357NJ0000F_0_0",
      "_LF3_",
      Array [
        "text",
        "LED",
        Array [
          "at",
          -4.064,
          1.778,
          0,
        ],
        Array [
          "effects",
          Array [
            "font",
            Array [
              "size",
              1.27,
              1.27,
            ],
          ],
          null,
          null,
          Array [
            "justify",
            "left",
          ],
          null,
        ],
      ],
    ],
    "_LF2_",
    Array [
      "symbol",
      "PC357NJ0000F_0_1",
      "_LF3_",
      Array [
        "arc",
        Array [
          "start",
          4.851400000000006,
          1.27,
        ],
        Array [
          "end",
          4.8260000000000005,
          -1.27,
        ],
        Array [
          "radius",
          Array [
            "at",
            4.838700000000003,
            0,
          ],
          Array [
            "length",
            1.27,
          ],
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "none",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "circle",
        Array [
          "center",
          -8.255,
          4.445,
        ],
        Array [
          "radius",
          0.381,
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "none",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "rectangle",
        Array [
          "start",
          -7.62,
          4.064,
        ],
        Array [
          "end",
          0.762,
          -4.318,
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "none",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "rectangle",
        Array [
          "start",
          1.778,
          4.064,
        ],
        Array [
          "end",
          7.62,
          -4.318,
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "none",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "polyline",
        Array [
          "pts",
          Array [
            "xy",
            7.62,
            2.54,
          ],
          Array [
            "xy",
            5.08,
            0.762,
          ],
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "none",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "polyline",
        Array [
          "pts",
          Array [
            "xy",
            5.08,
            -0.762,
          ],
          Array [
            "xy",
            7.62,
            -2.54,
          ],
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "none",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "polyline",
        Array [
          "pts",
          Array [
            "xy",
            5.08,
            2.286,
          ],
          Array [
            "xy",
            5.08,
            -2.286,
          ],
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "none",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "polyline",
        Array [
          "pts",
          Array [
            "xy",
            7.62,
            -2.54,
          ],
          Array [
            "xy",
            6.8580000000000005,
            -1.27,
          ],
          Array [
            "xy",
            6.096,
            -2.286,
          ],
          Array [
            "xy",
            7.62,
            -2.54,
          ],
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "outline",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "polyline",
        Array [
          "pts",
          Array [
            "xy",
            -0.508,
            -2.032,
          ],
          Array [
            "xy",
            -1.016,
            -1.016,
          ],
          Array [
            "xy",
            -1.524,
            -1.524,
          ],
          Array [
            "xy",
            -0.508,
            -2.032,
          ],
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "outline",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "polyline",
        Array [
          "pts",
          Array [
            "xy",
            0.254,
            -1.016,
          ],
          Array [
            "xy",
            -0.254,
            0,
          ],
          Array [
            "xy",
            -0.762,
            -0.508,
          ],
          Array [
            "xy",
            0.254,
            -1.016,
          ],
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "outline",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "polyline",
        Array [
          "pts",
          Array [
            "xy",
            -2.286,
            -0.254,
          ],
          Array [
            "xy",
            -0.508,
            -2.032,
          ],
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "none",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "polyline",
        Array [
          "pts",
          Array [
            "xy",
            -1.524,
            0.762,
          ],
          Array [
            "xy",
            0.254,
            -1.016,
          ],
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "none",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "polyline",
        Array [
          "pts",
          Array [
            "xy",
            -2.286,
            -1.27,
          ],
          Array [
            "xy",
            -6.35,
            -1.27,
          ],
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "none",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "polyline",
        Array [
          "pts",
          Array [
            "xy",
            -5.8420000000000005,
            1.27,
          ],
          Array [
            "xy",
            -4.318,
            -1.27,
          ],
          Array [
            "xy",
            -2.54,
            1.27,
          ],
          Array [
            "xy",
            -5.8420000000000005,
            1.27,
          ],
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "outline",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "polyline",
        Array [
          "pts",
          Array [
            "xy",
            -7.62,
            2.54,
          ],
          Array [
            "xy",
            -4.318,
            2.54,
          ],
          Array [
            "xy",
            -4.318,
            -2.54,
          ],
          Array [
            "xy",
            -7.62,
            -2.54,
          ],
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "none",
          ],
        ],
      ],
      "_LF3_",
      Array [
        "polyline",
        Array [
          "pts",
          Array [
            "xy",
            -1.524,
            -3.048,
          ],
          Array [
            "xy",
            0,
            -3.81,
          ],
          Array [
            "xy",
            -1.27,
            -3.81,
          ],
          Array [
            "xy",
            -5.08,
            -3.81,
          ],
          Array [
            "xy",
            -3.81,
            -3.048,
          ],
          Array [
            "xy",
            -1.524,
            -3.048,
          ],
        ],
        Array [
          "stroke",
          Array [
            "width",
            0,
          ],
        ],
        Array [
          "fill",
          Array [
            "type",
            "none",
          ],
        ],
      ],
    ],
    "_LF2_",
    Array [
      "symbol",
      "PC357NJ0000F_1_1",
      "_LF3_",
      Array [
        "pin",
        "input",
        "clock",
        Array [
          "at",
          -12.7,
          2.54,
          0,
        ],
        Array [
          "length",
          5.08,
        ],
        "_LF4_",
        Array [
          "name",
          "A",
          Array [
            "effects",
            Array [
              "font",
              Array [
                "size",
                1.27,
                1.27,
              ],
            ],
            null,
            null,
            null,
            null,
          ],
        ],
        "_LF4_",
        Array [
          "number",
          "1",
          Array [
            "effects",
            Array [
              "font",
              Array [
                "size",
                1.27,
                1.27,
              ],
            ],
            null,
            null,
            null,
            null,
          ],
        ],
      ],
      "_LF3_",
      Array [
        "pin",
        "output",
        "inverted",
        Array [
          "at",
          -12.7,
          -2.54,
          0,
        ],
        Array [
          "length",
          5.08,
        ],
        "_LF4_",
        Array [
          "name",
          "B",
          Array [
            "effects",
            Array [
              "font",
              Array [
                "size",
                1.27,
                1.27,
              ],
            ],
            null,
            null,
            null,
            null,
          ],
        ],
        "_LF4_",
        Array [
          "number",
          "2",
          Array [
            "effects",
            Array [
              "font",
              Array [
                "size",
                1.27,
                1.27,
              ],
            ],
            null,
            null,
            null,
            null,
          ],
        ],
      ],
      "_LF3_",
      Array [
        "pin",
        "passive",
        "inverted_clock",
        Array [
          "at",
          12.7,
          2.54,
          180,
        ],
        Array [
          "length",
          5.08,
        ],
        "_LF4_",
        Array [
          "name",
          "D",
          Array [
            "effects",
            Array [
              "font",
              Array [
                "size",
                1.27,
                1.27,
              ],
            ],
            null,
            null,
            null,
            null,
          ],
        ],
        "_LF4_",
        Array [
          "number",
          "4",
          Array [
            "effects",
            Array [
              "font",
              Array [
                "size",
                1.27,
                1.27,
              ],
            ],
            null,
            null,
            null,
            null,
          ],
        ],
      ],
      "_LF3_",
      Array [
        "pin",
        "bidirectional",
        "line",
        Array [
          "at",
          12.7,
          -2.54,
          180,
        ],
        Array [
          "length",
          5.08,
        ],
        "_LF4_",
        Array [
          "name",
          "C",
          Array [
            "effects",
            Array [
              "font",
              Array [
                "size",
                1.27,
                1.27,
              ],
            ],
            null,
            null,
            null,
            null,
          ],
        ],
        "_LF4_",
        Array [
          "number",
          "3",
          Array [
            "effects",
            Array [
              "font",
              Array [
                "size",
                1.27,
                1.27,
              ],
            ],
            null,
            null,
            null,
            null,
          ],
        ],
      ],
    ],
  ],
]
`;
